www_src/pages/section-2/section-2.jsx

```diff
@@ -1,14 +1,210 @@
+    }
+
+    return (
+      <div
+        className="page"
+        style={style}>
+      </div>
+    );
+  }
+});
+
+var AddPage = React.createClass({
+  render: function () {
+    return (
+      <button
+        onClick={this.handleClick}
+        className="add-page">
+      </button>
+    )
+  },
+  handleClick: function (event) {
+    this.props.onClick(event);
+  }
+})
+
+var Slot = React.createClass({
+  render: function () {
+    return (
+      <div
+        className="slot"
+        style={{
+          width: 100 / parseInt(this.props.perRow, 10) + '%',
+          height: 100 / parseInt(this.props.perRow, 10) + '%'
+        }}>
+          {this.props.children}
+      </div>
+    );
+  }
+});
+
+var Grid = React.createClass({
+  zoom: function (amount) {
+    this.setState({
+      zoom: amount
+    });
+  },
+  getInitialState: function () {
+    var width = Math.ceil(Math.random() * 16);
+    var height = Math.ceil(Math.random() * 16);
+
+    return {
+      zoom: this.props.initialZoom,
+      layout: generateLayout(width, height)
+    }
+  },
+  addPageClick: function (event) {
+    var newLayout = this.state.layout;
+
+    newLayout[event.y][event.x] = 'EMPTY';
+
+    this.setState({
+      layout: newLayout
+    });
+  },
+  render: function () {
+    var nodes = [];
+
+    var self = this;
+
+    // Determine if a slot has a neighboring page in any cardinal direction
+    function hasNeighbors(x, y) {
+      if (x > 0 && self.state.layout[y][x - 1]) {
+        return true;
+      }
+
+      if (x < self.state.layout[0].length - 1 && self.state.layout[y][x + 1]) {
+        return true;
+      }
+
+      if (y > 0 && self.state.layout[y - 1][x]) {
+        return true;
+      }
+
+      if (y < self.state.layout.length - 1 && self.state.layout[y + 1][x]) {
+        return true;
+      }
+
+      return false;
+    }
+
+    for (var y = 0; y < this.state.layout.length; y++) {
+      for (var x = 0; x < this.state.layout[0].length; x++) {
+        if (this.state.layout[y][x]) {
+          nodes.push(
+            <Slot x={x} y={y} perRow={this.state.layout[0].length}>
+              <Page screenshot={this.state.layout[y][x]} />
+            </Slot>
+          );
+        } else if (hasNeighbors(x, y)) {
+          nodes.push(
+            <Slot x={x} y={y} perRow={this.state.layout[0].length}>
+              {/* Overriding default click param to provide x/y coords without AddPage knowing them. */}
+              <AddPage onClick={ this.addPageClick.bind(this, {x:x, y:y}) }/>
+            </Slot>
+          );
+        } else {
+          nodes.push(
+            <Slot x={x} y={y} perRow={this.state.layout[0].length}/>
+          );
+        }
+      }
+    }
+
+    var gridTransform = {
+      transform: 'scale(' + this.state.zoom + ')',
+      WebkitTransform: 'scale(' + this.state.zoom + ')'
+    }
+
+    return (
+      <div className="grid" style={gridTransform}>
+        {nodes}
```

Comment:
when you render lists of nodes you need to add a `key` property to each one or else React will display a warning. you can just use the index in the array